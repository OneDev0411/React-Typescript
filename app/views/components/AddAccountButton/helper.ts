import IconRealtor from '@app/views/components/SvgIcons/Realtor/IconRealtor.svg'
import IconZillow from '@app/views/components/SvgIcons/Zillow/IconZillow.svg'

export const convertLeadChannelsToAuthAccounts = (
  channels: LeadChannel[]
): IOAuthAccount[] => {
  return channels.map(channel => ({
    brand: channel.brand,
    contacts_last_sync_at: (channel.last_capture_date
      ? new Date(channel.last_capture_date)
      : new Date()
    ).toISOString(),
    created_at: new Date(channel.created_at).toISOString(),
    deleted_at: channel.deleted_at
      ? new Date(channel.deleted_at).toISOString()
      : null,
    display_name: channel.source_type,
    email: channel.user,
    first_name: '',
    gid: '',
    history_id: '',
    id: channel.id,
    last_name: '',
    last_sync_duration: 0,
    messages_total: 0,
    ord: '',
    photo: null,
    profile_image_url:
      channel.source_type === 'Zillow' ? IconZillow : IconRealtor,
    resource_name: channel.source_type,
    revoked: false,
    scope: [],
    scope_summary: [],
    threads_total: channel.capture_number,
    type: channel.source_type,
    updated_at: (channel.updated_at
      ? new Date(channel.updated_at)
      : new Date()
    ).toISOString(),
    user: channel.user,
    jobs: [
      {
        brand: channel.brand,
        deleted_at: channel.deleted_at
          ? new Date(channel.deleted_at).toISOString()
          : null,
        id: channel.id,
        google_credential: null,
        job_name: 'calendar',
        metadata: null,
        microsoft_credential: null,
        recurrence: true,
        resume_at: null,
        start_at: new Date(channel.created_at).toISOString(),
        status: 'success',
        type: 'users_jobs',
        updated_at: (channel.updated_at
          ? new Date(channel.updated_at)
          : new Date()
        ).toISOString(),
        user: channel.user
      },
      {
        brand: channel.brand,
        deleted_at: channel.deleted_at
          ? new Date(channel.deleted_at).toISOString()
          : null,
        id: channel.id,
        google_credential: null,
        job_name: 'contacts',
        metadata: null,
        microsoft_credential: null,
        recurrence: true,
        resume_at: null,
        start_at: new Date(channel.created_at).toISOString(),
        status: 'success',
        type: 'users_jobs',
        updated_at: (channel.updated_at
          ? new Date(channel.updated_at)
          : new Date()
        ).toISOString(),
        user: channel.user
      },
      {
        brand: channel.brand,
        deleted_at: channel.deleted_at
          ? new Date(channel.deleted_at).toISOString()
          : null,
        id: channel.id,
        google_credential: null,
        job_name: 'email',
        metadata: null,
        microsoft_credential: null,
        recurrence: true,
        resume_at: null,
        start_at: new Date(channel.created_at).toISOString(),
        status: 'success',
        type: 'users_jobs',
        updated_at: (channel.updated_at
          ? new Date(channel.updated_at)
          : new Date()
        ).toISOString(),
        user: channel.user
      }
    ]
  }))
}
