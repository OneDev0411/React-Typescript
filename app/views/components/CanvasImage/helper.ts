export function fitImageIntoCanvas(
  img: HTMLImageElement,
  w: number,
  h: number
): Parameters<CanvasRenderingContext2D['drawImage']> {
  let containerRatio = h / w
  let width = img.naturalWidth
  let height = img.naturalHeight
  let imgRatio = height / width

  if (imgRatio > containerRatio) {
    // image's height too big
    height = width * containerRatio
  } else {
    // image's width too big
    width = height / containerRatio
  }

  return [img, 0, 0, width, height, 0, 0, w, h]
}
