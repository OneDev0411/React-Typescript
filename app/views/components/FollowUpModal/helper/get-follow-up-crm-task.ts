import fecha from 'fecha'

import { EmailThreadEmail } from 'components/EmailThread/types'
import { initialValueGenerator } from 'components/EventDrawer/helpers/initial-value-generator'
import { normalizeAssociations } from 'views/utils/association-normalizers'

import { Props as FollowUpProps } from '../FollowUpModal'
import { FollowUpEmail, FollowUpEmailCampaign } from '../types'

// TODO: should drop specific email association and make it general
export function getFollowUpCrmTask(
  email: FollowUpEmail | undefined,
  dueDate: Date,
  user: IUser,
  dictionary?: FollowUpProps['dictionary']
) {
  if (!email) {
    // return undefined
    const title = getCrmTaskTitle(undefined, dictionary?.taskTitle)
    const description = getCrmTaskDescription(
      user.display_name,
      dueDate,
      dictionary?.taskDescription
    )

    return initialValueGenerator(
      user,
      [],
      dueDate,
      undefined,
      title,
      description
    )
  }

  if ('type' in email && email.type === 'email_campaign') {
    return getCrmTaskFromEmailCampaign(email, dueDate, user, dictionary)
  }

  return getCrmTaskFromEmailThreadEmail(
    email as EmailThreadEmail,
    dueDate,
    user
  )
}

function getCrmTaskFromEmailCampaign(
  email: FollowUpEmailCampaign,
  dueDate: Date,
  user: IUser,
  dictionary?: FollowUpProps['dictionary']
) {
  const owner = email.from.type === 'user' ? email.from : user
  const contactAssociations = email.recipients
    .filter(resp => resp.contact)
    .map(resp => ({
      contact: resp.contact,
      association_type: 'contact'
    }))
  const title = getCrmTaskTitle(
    contactAssociations[0]?.contact.display_name,
    dictionary?.taskTitle,
    email
  )
  const description = getCrmTaskDescription(
    owner.display_name,
    email.due_at * 1000,
    dictionary?.taskDescription,
    email
  )

  const values = initialValueGenerator(
    owner,
    normalizeAssociations([
      ...contactAssociations,
      {
        association_type: 'email',
        email
      }
    ]),
    dueDate,
    undefined,
    title,
    description
  )

  return values
}

function getCrmTaskFromEmailThreadEmail(
  email: EmailThreadEmail,
  dueDate: Date,
  user: IUser,
  dictionary?: FollowUpProps['dictionary']
) {
  const owner = user
  const contactAssociations =
    email.thread?.contacts?.map?.(contact => ({
      contact,
      association_type: 'contact'
    })) || []
  const title = getCrmTaskTitle(email.to[0], dictionary?.taskTitle)
  const description = getCrmTaskDescription(
    owner.display_name,
    email.date,
    dictionary?.taskDescription
  )

  const values = initialValueGenerator(
    owner,
    normalizeAssociations(contactAssociations),
    dueDate,
    undefined,
    title,
    description
  )

  return values
}

// Todo: change this helper to be more general now it just accept email
export function getCrmTaskTitle(
  item?: string,
  getter?: (item?: string) => string,
  email?: FollowUpEmailCampaign
) {
  if (getter) {
    return getter(item)
  }

  let title = 'Follow up'

  if (email) {
    title = 'Follow up Email'

    const recipientType: IEmailRecipientType =
      email.recipients[0]?.recipient_type

    switch (recipientType) {
      case 'Email':
        return `${title}: ${
          email.recipients[0].contact?.display_name || email.recipients[0].email
        }`
      case 'Tag':
        return `${title}: ${email.recipients[0].tag}'s Tag`
      case 'AllContacts':
        return `${title}: All Contacts`
      default:
        return title
    }
  }

  if (item) {
    return `${title}: ${item}`
  }

  return title
}

// Todo: change this helper to be more general now it just accept email
export function getCrmTaskDescription(
  item: string,
  dueDate: Date | number,
  getter?: (item: string, dueDate: Date | number) => string,
  email?: FollowUpEmailCampaign
) {
  if (getter) {
    return getter(item, dueDate)
  }

  return `This is a follow up reminder ${item} set in Rechat, on ${fecha.format(
    dueDate,
    'dddd MMMM Do, YYYY'
  )} for the attached email.`
}
