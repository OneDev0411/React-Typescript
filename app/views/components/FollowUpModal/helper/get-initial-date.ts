interface ReturnShape {
  todayTimestamp: number
  tomorrowTimestamp: number
  oneDayTimestamp: number
}

/**
 * return base date for follow up
 * @param initialDate
 */
export const getInitialDate = (initialDate: Date = new Date()): ReturnShape => {
  const oneDayTimestamp = 24 * 3600000
  const todayTimestamp = initialDate.getTime()
  const tomorrowTimestamp = todayTimestamp + oneDayTimestamp

  return { todayTimestamp, tomorrowTimestamp, oneDayTimestamp }
}
