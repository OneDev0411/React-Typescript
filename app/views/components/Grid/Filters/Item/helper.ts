export const getCurrentValues = (
  isActive: boolean,
  values: ILabelValue[]
): string | boolean => {
  if (!isActive && (!values || values.length === 0)) {
    return 'Missing value'
  }

  return Array.isArray(values) && values.map(item => item.label).join(' OR ')
}
