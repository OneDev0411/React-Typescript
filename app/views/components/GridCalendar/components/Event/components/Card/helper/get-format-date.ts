import { EventContentArg } from '@fullcalendar/react'
import fecha from 'fecha'

/**
 * return event style for using in full calendar
 * @param event
 * @param fullCalendarEventObject // the event object provided by full calendar
 */
export const getFormatDate = (
  event: ICalendarEvent,
  fullCalendarEventObject?: EventContentArg
): string => {
  const { all_day, recurring, timestamp, end_date } = event
  const startDate = new Date(Number(timestamp || 0) * 1000)
  const endDate = new Date(Number(end_date || 0) * 1000)

  const isDifferentDay =
    startDate.getDate() < endDate.getDate() ||
    startDate.getMonth() < endDate.getMonth()

  // reset start and end time if is a all_dat event since server send us a utc time for all_day events
  if (all_day) {
    const dummyStart = new Date(startDate)
    /*
    We're doing this because for some all-day recurring event which
    its year doesn't set and 1800 is set automatically by server
    and we're going to use the current range of our calendar
    */
    const startYear =
      recurring && fullCalendarEventObject
        ? fullCalendarEventObject.view.currentStart.getUTCFullYear()
        : dummyStart.getUTCFullYear()

    startDate.setHours(
      dummyStart.getUTCHours(),
      dummyStart.getUTCMinutes(),
      dummyStart.getUTCSeconds(),
      0
    )
    startDate.setFullYear(
      startYear,
      dummyStart.getUTCMonth(),
      dummyStart.getUTCDate()
    )

    if (end_date) {
      const dummyEnd = new Date(endDate)

      endDate.setHours(
        dummyEnd.getUTCHours(),
        dummyEnd.getUTCMinutes(),
        dummyEnd.getUTCSeconds(),
        0
      )
      endDate.setFullYear(
        dummyEnd.getUTCFullYear(),
        dummyEnd.getUTCMonth(),
        dummyEnd.getUTCDate()
      )
    }
  }

  const startFormatter = !all_day ? 'dddd, MMM D, hh:mm A' : 'dddd, MMM D'
  const endFormatter =
    !all_day && isDifferentDay ? 'dddd, MMM D, hh:mm A' : 'hh:mm A'

  const startDateFormat = fecha.format(startDate, startFormatter)
  const endDateFormat = fecha.format(endDate, endFormatter)

  return `${startDateFormat}${
    !all_day && end_date ? ` - ${endDateFormat}` : ''
  }`
}
