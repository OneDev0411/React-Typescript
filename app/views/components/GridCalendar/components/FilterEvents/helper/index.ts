/*
  this value provided from api doc. which is available in below.
  https://docs.rechat.com/testing/calendar.html#header-event-types
*/

export const CRM_OBJECT_TYPES = ['crm_task', 'crm_association', 'contact']

// possible values for event_type of a celebration event
export const CELEBRATION_OBJECT_TYPE = 'contact_attribute'
export const CELEBRATION_EVENT_TYPES = [
  'birthday',
  'child_birthday',
  'wedding_anniversary',
  'home_anniversary',
  'work_anniversary'
]

// possible values for event_type of a deal event
export const DEAL_OBJECT_TYPE = 'deal_context'
export const DEAL_EVENT_TYPES = [
  'list_date',
  'expiration_date',
  'contract_date',
  'inspection_date',
  'option_period',
  'financing_due',
  'title_due',
  't47_due',
  'closing_date',
  'possession_date',
  'lease_executed',
  'lease_application_date',
  'lease_begin',
  'lease_end'
]
