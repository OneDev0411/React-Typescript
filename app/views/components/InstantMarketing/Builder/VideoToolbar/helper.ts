import { IKeyframe } from './types'

export function fixLeadingZero(value: number): string {
  const roundedValue = Math.floor(value)

  return roundedValue < 10
    ? `0${roundedValue.toFixed(0)}`
    : roundedValue.toFixed(0)
}

export function msToMinutesAndSeconds(ms: number): string {
  const minutes = Math.floor(ms / 60000)
  const seconds = (ms % 60000) / 1000

  return `${fixLeadingZero(minutes)}:${fixLeadingZero(seconds)}`
}

export function getSlotWidth(
  index: number,
  keyframes: IKeyframe[],
  duration: number
): number {
  const currentSlotStartTime = keyframes[index].at
  const nextSlotStartTime = keyframes[index + 1]?.at ?? duration
  const slotDuration = nextSlotStartTime - currentSlotStartTime

  return (slotDuration / duration) * 100
}

export function getSlotProgress(
  index: number,
  keyframes: IKeyframe[],
  currentTime: number,
  duration: number
): number {
  const currentSlotStartTime = keyframes[index].at
  const nextSlotStartTime = keyframes[index + 1]?.at ?? duration
  const slotDuration = nextSlotStartTime - currentSlotStartTime

  if (currentTime <= currentSlotStartTime) {
    return 0
  }

  if (currentTime >= nextSlotStartTime) {
    return 100
  }

  return ((currentTime - currentSlotStartTime) / slotDuration) * 100
}

export function normalizeKeyframes(
  keyframes: IKeyframe[],
  duration: number
): IKeyframe[] {
  const normalizedKeyframes = [
    ...(keyframes[0].at > 0 ? [{ at: 0 }] : []),
    ...keyframes
  ]

  return normalizedKeyframes.filter(keyframe => keyframe.at < duration)
}
