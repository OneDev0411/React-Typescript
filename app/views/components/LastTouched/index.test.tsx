import React from 'react'

import { render } from 'enzyme'
import toJson from 'enzyme-to-json'

import { frequencyToString } from '@app/components/Pages/Dashboard/Contacts/components/ManageRelationship/helper'
import fullContact from 'fixtures/contacts/full-contact.json'
import mockDate, { RealDate } from 'utils/test-utils/mock-date'

import { TestBed } from '../../../../tests/unit/TestBed'
import { AppTheme } from '../../../AppTheme'

import { LastTouched } from '.'

describe('Contact profile last touched component', () => {
  beforeEach(() => {
    mockDate('2019-05-26T12:04:04Z')
  })

  afterEach(() => {
    global.Date = RealDate
  })

  it('renders', () => {
    const wrapper = render(
      <TestBed>
        <AppTheme>
          <LastTouched contact={fullContact as any} />
        </AppTheme>
      </TestBed>
    )

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders touch periods without decimal points', () => {
    const wrapper = render(
      <TestBed>
        <AppTheme>
          <LastTouched contact={fullContact as any} />
        </AppTheme>
      </TestBed>
    )

    const decimalPointedPeriod =
      (fullContact.next_touch! - fullContact.last_touch!) / 86400

    expect(wrapper.text().includes(decimalPointedPeriod.toString())).toBeFalsy()
  })

  it('renders monthly touch reminder freq', () => {
    const wrapper = render(
      <TestBed>
        <AppTheme>
          <LastTouched contact={fullContact as any} />
        </AppTheme>
      </TestBed>
    )

    const freqString = frequencyToString(
      fullContact.touch_freq as any
    ).toLowerCase()

    expect(wrapper.text().includes(freqString)).toBeTruthy()
  })
})
