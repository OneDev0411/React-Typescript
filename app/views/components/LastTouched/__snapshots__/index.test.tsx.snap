// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact profile last touched component renders 1`] = `
<div
  class="MuiBox-root-3 MuiBox-root-4 makeStyles-wrapper-1"
  data-tour-id="last-touch"
>
  <span>
    Last Touch was 
    <b>
      just now
    </b>
  </span>
  <span>
    , you wanted to be in touch 
    <b>
      monthly
    </b>
  </span>
</div>
`;
